staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -52,17 +52,11 @@ const (
 	// after receiving a 'too high resource version' error.
 	resourceVersionTooHighRetrySeconds = 1
 
-	// eventFreshDuration is time duration of events we want to keep.
-	// We set it to `defaultBookmarkFrequency` plus epsilon to maximize
-	// chances that last bookmark was sent within kept history, at the
-	// same time, minimizing the needed memory usage.
-	eventFreshDuration = 75 * time.Second
-
 	// defaultLowerBoundCapacity is a default value for event cache capacity's lower bound.
 	// TODO: Figure out, to what value we can decreased it.
 	defaultLowerBoundCapacity = 100
 
-	// defaultUpperBoundCapacity  should be able to keep eventFreshDuration of history.
+	// defaultUpperBoundCapacity should be able to keep the required history.
```

Comment:
how does this relate to `eventFreshDuration`? Is it required to suffice for that duration? Or recommended? What happens if it is not enough?

Reply:
`defaultUpperBoundCapacity` is the max capacity of watch events within the watch cache for a single resource and AFAIK this is hardcoded right now.

With the current max capacity of the watch cache (`defaultUpperBoundCapacity` = 100\*1024) and the default value for `eventFreshDuration` (75s), we would need to generate over 100\*1024/75 ~ 1.35k QPS mutating churn within a single resource for 75s straight in order to hit a similar problem that we wanted to address with this PR. This QPS is waaay above what system components can produce in a default configuration of Kubernetes (100 QPS for kube-controller-manager and kube-scheduler) in the worst case scenario.

For higher values of `eventFreshDuration`, this QPS threshold becomes proportionally lower, though such QPS would have to persist for a proportionally longer period of time as well.

However, this still might become a bottleneck if one lifts the QPSes for these two system components or host their own bursty controller that produces a considerable churn on a particular resource.

I'm leaning towards exposing the watch cache's "upper bound capacity" as a configurable flag but in a subsequent PR to prevent this change's scope creep - WDYT?

Reply:
Would delegate the question to @serathius.

In any case, having this thought process he a comment wouldn't harm.

Reply:
`resizeCacheLocked` is the method where we adjust the watchcache's size:

https://github.com/kubernetes/kubernetes/blob/3ec97a445f036a38bfec6291dee661954138bac9/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go#L318-L323

In a subsequent change that exposes the max capacity of the watchcache, we could add some logging that might indicate that `w.upperBoundCapacity < w.capacity*2` - though we'd need to take care there since this might be a spammy log given this is called on every incoming watch event.